pgx/src/datum/array.rs

```diff
@@ -69,6 +71,33 @@ impl<'a, T: FromDatum + serde::Serialize> serde::Serialize for Array<'a, T> {
     }
 }
 
+impl<'a, T: FromDatum> Drop for Array<'a, T> {
+    fn drop(&mut self) {
+        if let Array {
+            ptr,
+            raw,
+            datum_palloc: Some(data),
+            elem_slice,
+            ..
+        } = self
+        {
+            // It's just a slice, dropping it doesn't "do" anything, but out of an abundance of caution:
+            mem::drop(elem_slice);
+            // No conflict possible with ditching this.
+            unsafe { pg_sys::pfree(data.as_ptr().cast()) };
```

Comment:
While dropping the Array, check if the `datum_palloc` field is Some, and then if so, drop the `elem_slice` (it's a no-op, but it means there's no other pointers) and `pfree` the pointer to the allocation of Datums.